library/include/borealis/core/touch/gesture_recognizer.hpp

```diff
@@ -0,0 +1,63 @@
+/*
+    Copyright 2021 XITRIX
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#pragma once
+
+#include <functional>
+
+#include <borealis/core/input.hpp>
+
+namespace brls
+{
+
+class View;
+
+// Represents current gesture state
+enum class GestureState
+{
+    INTERRUPTED, // Gesture has been interupted, no callbacks will come
+    UNSURE, // Gesture started recognition and not sure if it should interupt other recognizers
+    START, // Gesture sure that it match to conditions and will interupt other recognizers
+    STAY, // Gesture in process, user still hold finger on screen
+    END, // User released it's finger from screen, final frame of gesture
+    FAILED, // Gesture failed conditions
+};
+
+// Superclass for all the other recognizers
+class GestureRecognizer
+{
+public:
+    // Main recognition loop, for internal usage only, should not be called anywhere, but Application
+    virtual GestureState recognitionLoop(TouchState touch, View* view);
+    
+    // Returns true if recognizer plays default 'click' sound on touch start 
+    virtual bool soundOnTouch();
```

Comment:
I don't like storing the "should play a sound" state in a class attribute and have it set in recognitionLoop then returned in soundOnTouch.

IMO the "should a sound be played" boolean should be returned directly from recognitionLoop, and not carried around in the recognizer state like that.

I think it's going to be hard to return the boolean from recognitionLoop since you already return a GestureState, but it could be passed as a pointer?

`virtual GestureState recognitionLoop(TouchState touch, View* view, bool* shouldPlaySound);`

This whole sound situation could be improved a bit more actually, since we need to play the "nothing to touch here" sound when the user touches somewhere on the screen that doesn't have anything. A boolean might not be enough to represent all of our use cases.

So maybe instead of a "should play sound" boolean, we should add a "touch state" value that represents the gesture state: did it hit? did it miss? You go from view to view gathering the state, and then you play the sound in Application based on the final state.

I don't know if I'm really clear, let me know if you see what I mean

Reply:
I've made some changes in recognition loop, check it. I am fully agreed that something should be changed with "sound for nothing", but right now I have no any idea of how I could improve it...

"touch state" could be difficult to use, cause there are some cases where it's useless, i.e. pan recognizer, it's ok with touch, but it's still should play "nothing" sound, or tap recognizer should not play it, cause it has it's own "on pressed" sound, but if button is disabled, or has no actions on it, it should play "nothing" sound anyway.

Reply:
How about replacing the bool with the sound to play? And then use `brls::SOUND_NONE` if you don't want to play anything.